dali/core/mm/default_resources.cc

```diff
@@ -174,6 +177,7 @@ const std::shared_ptr<device_async_resource> &ShareDefaultDeviceResourceImpl(int
     CUDA_CALL(cudaGetDevice(&device_id));
   }
   if (g_resources.device.empty()) {
+    std::lock_guard<std::mutex> lock(g_resources.mtx);
     int ndevs = 0;
```

Comment:
Shouldn't you check `g_resources.device.empty()` again after taking the lock or we are fine with multiple `g_resources.device.resize(ndevs);` calls?

Reply:
Not really - cudaGetDeviceCount is a cheap call and if the array it's not empty, resize will be a no-op.